compiler/rustc_hir/src/def.rs

```diff
@@ -405,7 +405,7 @@ pub enum Res<Id = hir::HirId> {
     /// A tool attribute module; e.g., the `rustfmt` in `#[rustfmt::skip]`.
     ///
     /// **Belongs to the type namespace.**
-    ToolMod,
+    ToolMod { allow_shadow: bool },
```

Comment:
```suggestion
    ToolMod,
```
This feature should not introduce special ad hoc rules on shadowing like this.
I specifically asked on the RFC whether `diagnostic::foo` should be tool attributes and the answer was yes, so I don't think it should additionally bend language rules to serve its specific case, it already barely pulls its weight.

Reply:
This is somewhat required to workaround https://github.com/rust-lang/rust/issues/111135

To cite from the [tool attribute RFC](https://rust-lang.github.io/rfcs/2103-tool-attributes.html):

> During macro expansion, when faced with an attribute, the compiler first tries to find a macro using the macro name resolution rules. The compiler then checks if the attribute matches any of the declared or built- in attributes. 

Which means at least as far as I understand it's totally fine to allow shadowing there. 

I've chosen to make this shadowing behaviour optional and use it only for the new namespace for now to minimize possible breakage there. If you feel that's the wrong solution please suggest a better way to fix the underlying bug.

> I don't think it should additionally bend language rules to serve its specific case, it already barely pulls its weight.

I do not think that's the correct place to discuss whether or not the namespace pulls in its weight. It seems like that the language team feels that it pulls in it's weight, so you might want to raise that concern there.

Reply:
>The compiler then checks if the attribute matches any of the declared or built- in attributes.

Then this part is the random ad hoc language bending that should not be in the RFC.

>If you feel that's the wrong solution please suggest a better way to fix the underlying bug.

I think the better way is to remove the artificial errors from https://github.com/rust-lang/rust/issues/56414.
That issue says "this restriction can be lifted fully or partially if it causes too much trouble in practice with ambiguity errors, or with addition of new names to libraries being backward-incompatible due to new ambiguities", and it did consistently caused at least some amount of trouble, with `diagnostic` being just a single instance of that.

Reply:
> Then this part is the random ad hoc language bending that should not be in the RFC.

I don't think it's worth to argue about that. It is accepted stable language behaviour since 2016 according to RFC-2103 (tool attributes).

> I think the better way is to remove the artificial errors from #56414.
That issue says "this restriction can be lifted fully or partially if it causes too much trouble in practice with ambiguity errors, or with addition of new names to libraries being backward-incompatible due to new ambiguities", and it did consistently caused at least some amount of trouble, with diagnostic being just a single instance of that.

Just to be sure that I understand that correctly: You basically suggest to remove the `allow_shadow` flag and to suppress unconditionally the ambiguity error [here](https://github.com/rust-lang/rust/pull/111780/files#diff-c046507afdba3b0705638f53fffa156cbad72ed17aa01d96d7bd1cc10b8d9bceR639)?

Reply:
>I don't think it's worth to argue about that. It is accepted stable language behaviour since 2016 according to RFC-2103 (tool attributes).

Ah, that's from the tool attribute RFC.
It's not the actually implemented stable language behavior.
RFC details are typically tweaked based on implementation experience, and in this specific case the RFC text is especially outdated because it pre-dates macro modularization and uniform imports.
My suggestion is to follow the rules that are actually implemented since 2018 edition.

>Just to be sure that I understand that correctly: You basically suggest to remove the allow_shadow flag and to suppress unconditionally the ambiguity error [here](https://github.com/rust-lang/rust/pull/111780/files#diff-c046507afdba3b0705638f53fffa156cbad72ed17aa01d96d7bd1cc10b8d9bceR639)?

I need to check what exactly needs to be removed to fix https://github.com/rust-lang/rust/issues/56414, probably on Monday.

Reply:
I implemented the fix for https://github.com/rust-lang/rust/issues/56414 in https://github.com/rust-lang/rust/pull/112086.

Reply:
Thanks for putting up a fix that fast. i will rebase my PR on top of yours and remove the special handling in the next few days.